### PR TITLE
Refactor prompt_parser

### PR DESCRIPTION
**Pull request by ritugala**
Refactoring Instruction Parser Code

# Description

Refactored instruction parser code so that:
An OpenAI response which is a JSON string can be checked for multiple keys rather than just "instruction" and "demonstrations" ---  this will be useful for the column selection PR
Also included functionality for optional keys which will be used in the actual column selection part 

# References

https://github.com/neulab/prompt2model/issues/285
-Added the Union[ openai.ChatCompletion, Exception ] because I was getting an error in CI that it was not compatible with Python 3.9 - and I wasn't sure how else to fix it.
- Not sure if the unit files require test of their own or if prompt_parser tests are enough.

# Blocked by

- NA

### CODE REVIEW

File: prompt2model/utils/parse_json_responses.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+import openai
+
+from prompt2model.utils import api_tools, get_formatted_logger
+from prompt2model.utils.api_tools import API_ERRORS, handle_api_error
+
+logger = get_formatted_logger("ParseJsonResponses")
+
+
+class JsonParsingFromLLMResponse:
```
Comment (neubig) [inline prompt2model/utils/parse_json_responses.py:13]: I feel like this doesn't need to be a class. Could we just use module-level functions and pass in `max_api_calls` to `extract_response` and `parse_prompt_to_fields`?

File: prompt2model/utils/parse_json_responses.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            Parsed Response or throws error
+        """
+        chat_api = api_tools.default_api_agent
+        last_error = None
+        while True:
+            self.api_call_counter += 1
+            try:
+                response: Union[
+                    openai.ChatCompletion, Exception
+                ] = chat_api.generate_one_completion(
```
Comment (neubig) [inline prompt2model/utils/parse_json_responses.py:84]: Change to the following
```suggestion
                response: openai.ChatCompletion | Exception = chat_api.generate_one_completion(
```

and add

```python
from __future__ import annotations
```

to the imports section of the file.

File: prompt2model/prompt_parser/instr_parser.py [hunk 3]
```diff
@@ -84,31 +45,12 @@ def parse_from_prompt(self, prompt: str) -> None:
...
-                last_error = e
-                handle_api_error(e)
-
-            if self.max_api_calls and self.api_call_counter >= self.max_api_calls:
-                # In case we reach maximum number of API calls, we raise an error.
-                logger.error("Maximum number of API calls reached.")
-                raise RuntimeError(
-                    "Maximum number of API calls reached."
-                ) from last_error
+        json_parser = parse_json_responses.JsonParsingFromLLMResponse(
```
Comment (neubig) [inline prompt2model/prompt_parser/instr_parser.py:48]: Per comment below, change to function call.

**Review by neubig — CHANGES_REQUESTED** [review]
Thanks for this @ritugala ! I have a few small suggestions.

Also, for the PR name it'd be good to use a somewhat more descriptive name like "Refactor prompt_parser" or "Move prompt parser functions into utils"

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -52,8 +52,8 @@ repos:
...
-  - repo: https://github.com/DavidAnson/markdownlint-cli2
-    rev: v0.5.1
-    hooks:
-    - id: markdownlint-cli2
-    - id: markdownlint-cli2-fix
+  # - repo: https://github.com/DavidAnson/markdownlint-cli2
+  #   rev: v0.5.1
+  #   hooks:
+  #   - id: markdownlint-cli2
+  #   - id: markdownlint-cli2-fix
```
Comment (neubig) [inline .pre-commit-config.yaml:59]:
```suggestion
  - repo: https://github.com/DavidAnson/markdownlint-cli2
    rev: v0.5.1
    hooks:
    - id: markdownlint-cli2
    - id: markdownlint-cli2-fix
```

File: prompt2model/utils/parse_json_responses.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+from prompt2model.utils import api_tools, get_formatted_logger
+from prompt2model.utils.api_tools import API_ERRORS, handle_api_error
+
+logger = get_formatted_logger("ParseJsonResponses")
+
+
+def extract_response(
+    response: openai.Completion, required_keys: list, optional_keys: list
+) -> dict:
```
Comment (neubig) [inline prompt2model/utils/parse_json_responses.py:16]:
```suggestion
) -> dict | None:
```

File: prompt2model/utils/parse_json_responses.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+            response: openai.ChatCompletion | Exception = (
+                chat_api.generate_one_completion(
+                    prompt,
+                    temperature=0.01,
+                    presence_penalty=0,
+                    frequency_penalty=0,
+                )
+            )
+            extraction = extract_response(response, required_keys, optional_keys)
+            if extraction != {}:
```
Comment (neubig) [inline prompt2model/utils/parse_json_responses.py:87]:
```suggestion
            if extraction is not None:
```

File: prompt2model/utils/parse_json_responses.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        required and optional keys then returns the
+        final response as a Dictionary
+        Else returns None.
+    """
+    response_text = response.choices[0]["message"]["content"]
+    try:
+        response_json = json.loads(response_text, strict=False)
+    except json.decoder.JSONDecodeError:
+        logger.warning(f"API response was not a valid JSON: {response_text}")
+        return {}
```
Comment (neubig) [inline prompt2model/utils/parse_json_responses.py:35]:
```suggestion
        return None
```

File: prompt2model/utils/parse_json_responses.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+    missing_keys = [key for key in required_keys if key not in response_json]
+    if len(missing_keys) != 0:
+        logger.warning(f'API response must contain {", ".join(required_keys)} keys')
+        return {}
```
Comment (neubig) [inline prompt2model/utils/parse_json_responses.py:40]:
```suggestion
        return None
```

**Review by neubig — CHANGES_REQUESTED** [review]
This looks good! I just have two small suggested changes, which can probably just be done by using Github's "accept suggestion" functionality:

- Revert the commenting of markdownlint
- Return None (as in the original code) instead of the empty accuracy

**Comment by ritugala** [comment]
Thank you for the changes! Sorry I didn't noticed you had left additional commits till after I pushed in the new commits - and sorry if you got too many notifications from all this!

**Review by ritugala — COMMENTED** [review]
Added the final changes

**Review by neubig — APPROVED** [review]

**Pull request merged**